### ISSUE TITLE
White level from camconst.json is not used for hdr dng files

### ISSUE THREAD

**Issue by Beep6581** [Beep6581/RawTherapee#2761]
Originally reported on Google Code with ID 2781

```
Create an hdr dng from an image with overexposed sky (clouds).
Apply neutral profile to the original raw file and to the dng file in rt.
dng file shows pinkish clouds, while original file is ok.

Samples uploaded to http://filebin.net/ytq6krp0c2

Ingo
```

Reported by `heckflosse@i-weyrich.de` on 2015-05-15 13:27:12

**Comment by Beep6581** [comment]
@heckflosse @iliasg status?

**Comment by heckflosse** [comment]
@Beep6581 As a workaround I added possibility to specify custom white level in hdrmerge some time ago. That `fixed` the problem.

**Comment by iliasg** [comment]
Hmm .. what if .. at the raw tab, we had a GUI puldown list to alter the WL priorities between DNG_exif and camconst.json ?

**Comment by Beep6581** [comment]
Would it not be best if the user ran exiv2/exiftool or any of the available GUIs to fix the white levels in their files?

Which files other than HDR DNG does this issue address?

**Comment by iliasg** [comment]
@Beep6581 your proposal is not best for the normal users (little to zero knowledge about white level, exiftool syntax)

I think all third party DNGs can possibly be affected (HDRmerge, magic Lantern, custom made) even Adobe's at early support stages (I remember a case with Panasonic LX100  @heckflosse ;) )

Alternative option is to use a paragraph in RTpedia for exactly this issue (problems from non correct white level data in DNG exif) and suggest playing with raw.white point correction or exitool corrections taking data from camconst.json (but what if there are no data in camconst.json ??).

**Comment by heckflosse** [comment]
@maybe an easy solution for the case the dng is produced by hdrmerge would be to use camconst for floating point dng files...

**Comment by iliasg** [comment]
@heckflosse Alex @ Magic Lantern also exports float_DNG with a program like HDRmerge .. CeroNoise
https://www.magiclantern.fm/forum/index.php?topic=9581.0
We could have a conflict if he uses a different approach for WL reporting in his float DNGs which is correct for his file and different from camconst.json ..

**Comment by heckflosse** [comment]
@iliasg Ok, then this is not an option.

**Comment by iliasg** [comment]
@heckflosse yes .. not an option .. maybe .. but ..

- CeroNoise is stalled and has very limited user base .. although it or any equivalent (Lumariver ?) can come up again :(

- If there is a signature of HDRMerge in the exif then use this info and act accordingly ;)

**Comment by Beep6581** [comment]
> your proposal is not best for the normal users (little to zero knowledge about white level, exiftool syntax)

@iliasg I agree with that. These same users with little to zero knowledge would come asking us for help, and we'd have to analyze their files and find the correct white levels to put into camconst.json so that the suggested combobox could be used. Or am I wrong? If I'm not wrong, then we are faced with two options - both require effort, but one offers a permanent fix (edit bad Exif) while the other is temporary (work-around in RT). Maybe coding a combobox and adding support for handling broken DNG files from multiple sources would be more work than writing a RawPedia article explaining how to find correct white levels with a link to a simple to use GUI Exif editor from which they could fix it, with CLI examples as well. Just a thought.

**Comment by iliasg** [comment]
@Beep6581 
The best fix (quality wise) I see is a patched HDRmerge (or equivalent) to make it use camconst.json data .. this will give reliable results from the start. The current state of HDRmerge after Ingo's HDRMerge patch (ability to use custom WL) is also at the right path .. 
Correcting merged data is not the same as doing the merge correctly. Think of the case where the merged files are not at same ISO ..
For other mergedDNG exporters .. we can ask the same .. may be we will find helping hands and brains this way ..

For RT I think a combobox or even a check box (use camconst data as reference for DNGs) to alter priority order will give freedom to find the best settings and as you say (I like your thought) push the users to supply samples .. ;)

I am many times in the case of needing a "priority" tag in camconst (instead of these ungly adobe_coeff workarounds in Dcraw.cc) .. not only for floatDNGs .. I have met Adobe DNGs with problems (LX100), Pentax in_camera DNGs with problematic WL (it is reported after BL subtraction ..) .. Leica DNGs with non standard color matrix (d50) .. many smartphones with strange color data and structure .. maybe others I haven't checked yet .. 
And there are and some other features missing from camconst (like multiple frame sizes ..) .. We have to review and update the camconst structure in the near future I think ..

**Comment by Beep6581** [comment]
@iliasg @heckflosse could you please update the title of this issue to better match the desired outcome, whatever that is? Seems like it's not a bug but an enhancement for a white level priority combobox to Raw > Raw White Levels?
